extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeDashManifestCreator.java

```diff
@@ -0,0 +1,1887 @@
+     *         status code 404 after redirects, and if the URL is valid);</li>
+     *         <li>an {@link ItagItem}, which needs to contain the following information:
+     *              <ul>
+     *                  <li>its type (see {@link ItagItem.ItagType}), to identify if the content is
+     *                  an audio or a video stream;</li>
+     *                  <li>its bitrate;</li>
+     *                  <li>its mime type;</li>
+     *                  <li>its codec(s);</li>
+     *                  <li>for an audio stream: its audio channels;</li>
+     *                  <li>for a video stream: its width and height.</li>
+     *              </ul>
+     *         </li>
+     *         <li>the duration of the video, which will be used if the duration could not be
+     *         parsed from the first sequence of the stream.</li>
+     *     </ul>
+     * </p>
+     *
+     * <p>In order to generate the DASH manifest, this method will:
+     *      <ul>
+     *          <li>request the first sequence of the stream (the base URL on which the first
+     *          sequence parameters are appended (see {@link #RN_0} and {@link #SQ_0})) with a POST
+     *          or GET request (depending of the client on which the streaming URL comes from);
+     *          </li>
+     *          <li>follow its redirection(s), if any;</li>
+     *          <li>save the last URL, remove the first sequence parameters;</li>
+     *          <li>use the information provided in the {@link ItagItem} to generate all
+     *          elements of the DASH manifest.</li>
+     *      </ul>
+     * </p>
+     *
+     * <p>
+     * If the duration cannot be extracted, the {@code durationSecondsFallback} value will be used
+     * as the stream duration.
+     * </p>
+     *
+     * @param postLiveStreamDvrStreamingUrl the base URL of the post-live-DVR stream/ended
+     *                                      livestream, which cannot be null
+     * @param itagItem                      the {@link ItagItem} corresponding to the stream, which
+     *                                      cannot be null
+     * @param targetDurationSec             the target duration of each sequence, in seconds (this
+     *                                      value is returned with the targetDurationSec field for
+     *                                      each stream in YouTube player response)
+     * @param durationSecondsFallback       the duration of the ended livestream which will be used
+     *                                      if the duration could not be extracted from the first
+     *                                      sequence
+     * @return the manifest generated into a string
+     * @throws YoutubeDashManifestCreationException if something goes wrong when trying to generate
+     *                                              the DASH manifest
+     */
+    @Nonnull
+    public static String createDashManifestFromPostLiveStreamDvrStreamingUrl(
```

Comment:
It must not be not removed at all since it allow the generation of post live sources, and we can't test this, as explained in the `StreamType.POST_LIVE_STREAM` enum (streams are rencoded later as normal streams by YouTube, so we can't use a specific video and we can't really mock responses because we need to do requests to video and audio streams URLs (some information needs to be extracted from the headers of the response)).

In the extractor yes, it is not used, but it will be used by clients, mostly NewPipe :)

Reply:
>  we can't test this, as explained in the StreamType.POST_LIVE_STREAM enum (streams are rencoded later as normal streams by YouTube, so we can't use a specific video and we can't really mock responses because we need to do requests to video and audio streams URLs (some information needs to be extracted from the headers of the response))

Short question: Why do we even need it then?

If you watch a live-stream you get a LIVE_STREAM stream type. If you watch the video afterwards you get a VIDEO_STREAM. 

I wouldn't implement this at all because that's such a corner case.... Not many people will ever encounter this situation and the code we have to adjust and maintain is huge. After the LIVE_STREAM stopped we could also be able simply reload the player and then get the VIDEO_STREAM.
Update: I found exactly ONE issue with 0 👍 → https://github.com/TeamNewPipe/NewPipe/issues/6232

I also noticed that POST_LIVE_AUDIO_STREAM is never used, so not sure why this also exists...

IMHO we should
* either remove POST_LIVE_STREAM and POST_LIVE_AUDIO_STREAM at all
* or at least - if they are platform specific (YT) - encapsulate them inside the platform specific code

Reply:
> If you watch the video afterwards you get a VIDEO_STREAM.
> [...]
> After the LIVE_STREAM stopped we could also be able simply reload the player and then get the VIDEO_STREAM.

Not just after, recently ended livestreams are considered as livestreams, because YouTube takes time to rencode them (most a few hours, but sometimes, it can take even a day)! That's the goal of the POST_LIVE_STREAM stream type.

> Short question: Why do we even need it then?

To play fully ended livestreams: instead of only being able to play a part of an ended livestream (always most of times 2 or 4 hours), you can play it fully (so if the stream was 7 hours long, you can play again 7 hours and not only the last 2 or 4 hours)!

In NewPipe, it also fixes the issue with the preview thumbnails, where the ones of the beginning are set at the available part in the HLS manifest used for playback (replaced by DASH manifests generated by `YoutubeDashManifestCreator`).

So no, we should not at all remove this improvement.

> I also noticed that POST_LIVE_AUDIO_STREAM is never used, so not sure why this also exists...

Who knows if an audio service, already available in the extractor or not, will have the same caracteristics as YouTube livestreams for live contents? That's why I added it as a part of the concept of stream types.

Reply:
@litetex Ping